### PR TITLE
feat: added CLI tool and overhauled documentation

### PR DESCRIPTION
**Pull request by Will-Mann-16**
Added a CLI tool utilising a caching mechanism, which is explained in teh documentation
Fixed some bugs that cropped up with testing, and writing examples (revolving make and reserved keywords)
Added `examples` directory, and `docs` directory with more information explaining different examples.

Sorry in advance, HUGE PR wanted to get in before AFK for a while.

### CODE REVIEW

File: blacklist.go
Comment (TommoLeedsy) [inline blacklist.go:1]: Can we rename this file and associated functions to use inclusive language, maybe allowlist instead?

File: docs/blacklisting.md
Comment (TommoLeedsy) [inline docs/blacklisting.md:1]: Same inclusive language issue

File: inputs/gin/parseFunction.go [hunk 1]
```diff
@@ -477,13 +483,15 @@ func parseFunction(s *gengo.Service, log zerolog.Logger, currRoute *gengo.Route,
 	return nil
 }
 
-func parseFromCalledFunction(log zerolog.Logger, callExpr *ast.CallExpr, argNo int, pkgName, pkgPath string, imports []*ast.ImportSpec, onExtract func(result astUtils.ParseResult)) (error, bool) {
+// parseFromCalledFunction parses a function call that an argument is a reference to something
+// It returns true if a new pass is needed to parse the function (i.e. it has had to extract another different type)
+func parseFromCalledFunction(s *gengo.Service, log zerolog.Logger, callExpr *ast.CallExpr, argNo int, pkgName, pkgPath, workDir string, imports []*ast.ImportSpec, onExtract func(result astUtils.ParseResult)) (error, bool) {
```
Comment (TommoLeedsy) [inline inputs/gin/parseFunction.go:488]: This function has unnecessary argument as they are contained in gengo.Service

Example of new function argument:

```func parseFromCalledFunction(s *gengo.Service, callExpr *ast.CallExpr, argNo int, pkgName, pkgPath string, imports []*ast.ImportSpec, onExtract func(result astUtils.ParseResult)) (error, bool)```

File: inputs/gin/parseFunction.go [hunk 2]
```diff
@@ -536,33 +585,77 @@ func parseFromCalledFunction(log zerolog.Logger, callExpr *ast.CallExpr, argNo i
...
 
-func parseIdentAndTrace(log zerolog.Logger, argType *ast.Ident, pkgPath string, pkgName string, imports []*ast.ImportSpec, onExtract func(result astUtils.ParseResult)) (error, bool) {
-	assignStmt, ok := argType.Obj.Decl.(*ast.AssignStmt)
-	if !ok {
-		return nil, false
-	}
+// parseIdentAndTrace parses an identifier and traces it back to its definition
+// It returns true if a new pass is needed to parse the function (i.e. it has had to extract another different type)
+// It is designed to match any number of arguments on either side
+func parseIdentAndTrace(s *gengo.Service, log zerolog.Logger, argType *ast.Ident, pkgPath, pkgName, workDir string, imports []*ast.ImportSpec, onExtract func(result astUtils.ParseResult)) (error, bool) {
```
Comment (TommoLeedsy) [inline inputs/gin/parseFunction.go:591]: Similarly, this function has unnecessary argument as they are contained in gengo.Service

Example of new function argument:

```func parseIdentAndTrace(s *gengo.Service, argType *ast.Ident, pkgPath, pkgName string, imports []*ast.ImportSpec, onExtract func(result astUtils.ParseResult)) (error, bool)```

File: inputs/gin/createRoutes.go
Comment (TommoLeedsy) [inline inputs/gin/createRoutes.go:1]: Same inclusive language issue

File: inputs/gin/parseFunction.go (hunk 1 above)
Comment (Will-Mann-16) [reply inputs/gin/parseFunction.go:488]: Agree with not doing `workDir`, but for now tabling the zerolog.Logger due to requirement for custom logger with custom variables.

Comment (Will-Mann-16) [reply inputs/gin/parseFunction.go:488]: Agree with not doing `workDir`, but for now tabling the zerolog.Logger due to requirement for custom logger with custom variables.

File: inputs/gin/parseFunction.go (hunk 2 above)
Comment (Will-Mann-16) [reply inputs/gin/parseFunction.go:591]: Same as before

**Pull request merged**